src/toast/observation_dist.py

```diff
@@ -164,170 +193,285 @@ def __init__(
-    newdist = DistDetSamp(
-        self._samples,
-        self._telescope.focalplane.detectors,
-        self._sample_sets,
-        self._detector_sets,
-        self._comm,
-        process_rows,
+    # Per-process send and receive information in terms of the flat
+    # packed buffers.
+    send_counts = AlignedI32(comm.size)
+    recv_counts = AlignedI32(comm.size)
+    send_displ = AlignedI32(comm.size)
+    recv_displ = AlignedI32(comm.size)
+
+    off = 0
+    send_shape = list()
+    for iproc, send_slc in enumerate(send_info):
+        send_displ[iproc] = off
+        n_elem = 0
+        if send_slc is not None:
+            n_elem = 1
+            ss = list()
+            for slc in send_slc:
+                slen = slc.stop - slc.start
+                n_elem *= slen
+                ss.append(slen)
+            send_shape.append(tuple(ss))
+        else:
+            send_shape.append(None)
+        send_counts[iproc] = n_elem
+        off += n_elem
+    send_n_elem = off
+
+    off = 0
+    recv_shape = list()
+    for iproc, recv_slc in enumerate(recv_info):
+        recv_displ[iproc] = off
+        n_elem = 0
+        if recv_slc is not None:
+            n_elem = 1
+            rs = list()
+            for slc in recv_slc:
+                rlen = slc.stop - slc.start
+                n_elem *= rlen
+                rs.append(rlen)
+            recv_shape.append(tuple(rs))
+        else:
+            recv_shape.append(None)
+        recv_counts[iproc] = n_elem
+        off += n_elem
+    recv_n_elem = off
+
+    # Set up the send / receive buffers
+    send_buf = buffer_class(send_n_elem)
+    recv_buf = buffer_class(recv_n_elem)
+
+    # Copy data into the send buffer
+    for iproc, send_slc in enumerate(send_info):
+        off = send_displ[iproc]
+        n_elem = send_counts[iproc]
+        if n_elem > 0:
+            send_buf[off : off + n_elem] = input[send_slc].flatten()
+
+    # Communicate data
+    comm.Alltoallv(
+        [
+            send_buf,
+            send_counts,
+            send_displ,
+            mpi_type,
+        ],
+        [
+            recv_buf,
+            recv_counts,
+            recv_displ,
+            mpi_type,
+        ],
     )
 
-    det_order = {y: x for x, y in enumerate(self.all_detectors)}
+    # Copy data to output
+    for iproc, recv_slc in enumerate(recv_info):
+        off = recv_displ[iproc]
+        n_elem = recv_counts[iproc]
+        if n_elem > 0:
+            output[recv_slc] = (
+                recv_buf[off : off + n_elem].array().reshape(recv_shape[iproc])
+            )
+
+    # Delete buffers
+    send_buf.clear()
+    del send_buf
+    recv_buf.clear()
+    del recv_buf
+    send_counts.clear()
+    del send_counts
+    send_displ.clear()
+    del send_displ
+    recv_counts.clear()
+    del recv_counts
+    recv_displ.clear()
+    del recv_displ
+
+
+def redistribute_data(old_dist, new_dist, shr_mgr, det_mgr, ivl_mgr, times=None):
```

Comment:
`shr_mgr, det_mgr, ivl_mgr` seem rather terse and cryptic for the input variable names.

Reply:
These are abbreviations for "`shared_manager`", "`detdata_manager`", and "`intervals_manager`".  I just hate typing excessively long variable names :-)  I can regex replace them if you like.

Reply:
This is obviously a matter of taste. I like the descriptive names better. `intervals_manager` feels like an improvement over `ivl_mgr`. Anyway, it is your call.

Reply:
Ok, I'll expand the variable names.  And since the classes themselves use the `Mgr` suffix, I'll change that suffix to `Manager` for consistency.